src/rgen/Microsoft.Macios.Bindings.Analyzer/NativeObjectHandleAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.Macios.Bindings.Analyzer;
+
+[DiagnosticAnalyzer (LanguageNames.CSharp)]
+public class NativeObjectHandleAnalyzer : DiagnosticAnalyzer {
+	internal static readonly DiagnosticDescriptor RBI0014 = new (
+		"RBI0014",
+		new LocalizableResourceString (nameof (Resources.RBI0014Title), Resources.ResourceManager, typeof (Resources)),
+		new LocalizableResourceString (nameof (Resources.RBI0014MessageFormat), Resources.ResourceManager,
+			typeof (Resources)),
+		"Usage",
+		DiagnosticSeverity.Error,
+		isEnabledByDefault: true,
+		description: new LocalizableResourceString (nameof (Resources.RBI0014Description), Resources.ResourceManager,
+			typeof (Resources))
+	);
+
+	public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (RBI0014);
+
+	public override void Initialize (AnalysisContext context)
+	{
+		context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.None);
+		context.EnableConcurrentExecution ();
+		context.RegisterSyntaxNodeAction (AnalyzeNode, SyntaxKind.SimpleMemberAccessExpression);
+	}
+
+	private bool IsINativeObject (ITypeSymbol typeSymbol)
+	{
+		return typeSymbol.AllInterfaces.Any (i => i.ToDisplayString () == "ObjCRuntime.INativeObject");
+	}
+
+	private bool IsSafeMethod (IMethodSymbol methodSymbol)
+	{
+		var name = methodSymbol.Name;
+		// White-list couple of methods that return one of the input parameters or a static
+		// object.
+		return name == "ThrowOnNull" || name == "DangerousAutorelease" || name == "GetConstant";
```

Comment:
Should we move this to a set, easier to maintain and probably will do less string comparisons.